.github/workflows/sel4test-hw.yml

```
@@ -0,0 +1,125 @@
+# Copyright 2021, Proofcraft Pty Ltd
+#
+# SPDX-License-Identifier: BSD-2-Clause
+
+# sel4test hardware builds and runs
+#
+# See sel4test-hw/builds.yml in the repo seL4/ci-actions for configs.
+
+name: seL4Test-HW
+
+on:
+  workflow_call:
+
+# intended to run on
+#  pull_request_target:
+#    types: [opened, reopened, synchronize, labeled]
+# needs PR target for secrets access; guard by requiring label
+
+# downgrade permissions to read-only as you would have in a standard PR action
+permissions:
+  contents: read
+
+jobs:
+  code:
+    name: Freeze Code
+    runs-on: ubuntu-latest
+    if: ${{ github.event_name == 'pull_request_target' &&
+            (
+              github.event.action != 'labeled' &&
+              (contains(github.event.pull_request.labels.*.name, 'hw-build') ||
+               contains(github.event.pull_request.labels.*.name, 'hw-test')) ||
+              github.event.action == 'labeled' &&
+              (github.event.label.name == 'hw-build' ||
+               github.event.label.name == 'hw-test')
+            ) }}
+    outputs:
+      xml: ${{ steps.repo.outputs.xml }}
+    steps:
+    - id: repo
+      uses: seL4/ci-actions/repo-checkout@master
+      with:
+        manifest_repo: sel4test-manifest
+        manifest: master.xml
+        sha: ${{ github.event.pull_request.head.sha }}
+
+  hw-build:
+    name: HW Build
+    runs-on: ubuntu-latest
+    needs: code
+    # To reduce the load (especially on the machine queue) we cancel any older
+    # runs of this workflow for the current PR.
+    concurrency:
+      group: ${{ github.workflow }}-sel4test-build-pr-${{ github.event.number }}-${{ strategy.job-index }}
+      cancel-in-progress: true
+    strategy:
+      fail-fast: false
+      matrix:
+        # There is no "rv32imac" hardware yet.
+        march: [armv7a, armv8a, nehalem, rv64imac]
+        compiler: [gcc, clang]
+    steps:
+    - name: Build
+      uses: seL4/ci-actions/sel4test-hw@master
+      with:
+        xml: ${{ needs.code.outputs.xml }}
+        march: ${{ matrix.march }}
+        compiler: ${{ matrix.compiler }}
+    - name: Upload images
+      uses: actions/upload-artifact@v4
+      with:
+        name: images-${{ matrix.march }}-${{ matrix.compiler }}
```

```suggestion
        name: sel4test-images-${{ matrix.march }}-${{ matrix.compiler }}
```
I have done some experimenting with running both `sel4test` and `sel4bench`. Since they are all part of the same workflow now, there can be upload conflicts here. 
- `sel4bench` uploads as `images-${{ matrix.march }}`
- `sel4test` uploads as `images-${{ matrix.march }}-${{ matrix.compiler }}`

---

No, I don't think that's a good idea, those names are partially used in other scripts that people have and are used to. I don't see any benefit in changing them, they are local to their specific workflow and make sense in their context.

---

We have only one shared worflow `sel4test-hw.yml` in the sel4 kernel project now that triggers both "sel4test" and "sel4bench" . So all these artifacts end up together, see e.g. the artifact list. At the moment with just these it works because the naming pattern is different. If another project is added here, it need to use a better pattern. And if we consider running "sel4bench" for the compilers it will also clash badly. Luckly the upload action warns on duplicate names now instead of silently overwriting them.

---

The upload action actually fails now if there is a duplicate. It's true that they will end up in the same summary screen. We'd still need to track down what depends on this, so we should not just change artifact names if we don't need to. If it does become a problem in the future, we can still change things then (and it's then only one central workflow, so it's much easier to do).

We don't need to get everything perfect now, we won't achieve that anyway. We should observe this in action for a while and then see what we need.

---

Ok, I can see how the UI will be confusing for the `images` part. I don't think those are used in automation outside this workflow (but kernel.elf might be). I'll try to double check on that first part and change those.